Support Apple MPS acceleration

ClaudiaComito opened the pull request:
**LAST EDITED MAR 18 2025**
[Note from human: 

**We extend PyTorch's MPS support to our users "as is". This is only tested on 1 MPI process, on the M1 chip.**

the most important changes in this PR are:
- Apple MPS are now a valid GPU device in `device.py`. Both `ht.array(..., device="gpu")` and `ht.array(..., device="mps")` are allowed. 
- `device`  attribute introduced for `ht.random.permutation`
- added an item to the PR template checklist to test with MPS (manually for now until the CI is expanded, see #1747 )
- codecov is unhappy but the introduced changes cannot be tested with our current setup.

Below a copilot summary]

This pull request includes various changes to add support for Metal Performance Shaders (MPS) on Apple devices and to handle specific cases where MPS does not support certain data types. The most important changes include updates to unit tests, skipping tests that fail on MPS, and modifications to core functionalities to ensure compatibility with MPS.

### MPS Support and Compatibility:

* [`.github/PULL_REQUEST_TEMPLATE.md`](diffhunk://#diff-18813c86948efc57e661623d7ba48ff94325c9b5421ec9177f724922dd553a35R9): Added a new checklist item for unit tests to ensure MPS compatibility.
* [`heat/core/devices.py`](diffhunk://#diff-e466c2ffdeb07e6f50d7b90825407b4d55a7aee44ca3ff900ac02f8cf48dd0ddL19-R25): Added support for MPS devices, including creating a new GPU device for Apple M1/M2 and updating device initialization and equality checks. [[1]](diffhunk://#diff-e466c2ffdeb07e6f50d7b90825407b4d55a7aee44ca3ff900ac02f8cf48dd0ddL19-R25) [[2]](diffhunk://#diff-e466c2ffdeb07e6f50d7b90825407b4d55a7aee44ca3ff900ac02f8cf48dd0ddR37-R38) [[3]](diffhunk://#diff-e466c2ffdeb07e6f50d7b90825407b4d55a7aee44ca3ff900ac02f8cf48dd0ddR138-R159)

### Unit Tests Adjustments:

* [`heat/cluster/tests/test_batchparallelclustering.py`](diffhunk://#diff-bba52bd905dd1cd60e86a18e43bf3c65abb8d6ac6becac74ffc8b442d50466ceR2): Added checks and skips for MPS in various test functions to handle unsupported cases. [[1]](diffhunk://#diff-bba52bd905dd1cd60e86a18e43bf3c65abb8d6ac6becac74ffc8b442d50466ceR2) [[2]](diffhunk://#diff-bba52bd905dd1cd60e86a18e43bf3c65abb8d6ac6becac74ffc8b442d50466ceR15-R20) [[3]](diffhunk://#diff-bba52bd905dd1cd60e86a18e43bf3c65abb8d6ac6becac74ffc8b442d50466ceR59) [[4]](diffhunk://#diff-bba52bd905dd1cd60e86a18e43bf3c65abb8d6ac6becac74ffc8b442d50466ceR94-R106)
* `heat/cluster/tests/test_kmeans.py`, `heat/cluster/tests/test_kmedians.py`, `heat/cluster/tests/test_kmedoids.py`: Adjusted data types used in tests to avoid unsupported float64 on MPS. [[1]](diffhunk://#diff-08bc033941cd21d47e68df088fe77e7cf576411a8c44c4a76a9b19a649e314dbR103-R116) [[2]](diffhunk://#diff-87d5072a85f048e98b01fdac6bd5e55d151b1695c1c263eda06353172d3efc2fR103-R109) [[3]](diffhunk://#diff-adf023d72d4dd16fadd96f07d20ac175a41bd0dba53645a36b322a1f5d0e1954R106-R112)

### Core Functionality Updates:

* [`heat/core/_operations.py`](diffhunk://#diff-ef0c46a534c74a4d5c6594bd14a6cf982c51f861ca18e8b2a8f2a42266e67540R200-R203): Added checks and warnings for unsupported float64 on MPS in various operations, including cumulative operations and local operations. [[1]](diffhunk://#diff-ef0c46a534c74a4d5c6594bd14a6cf982c51f861ca18e8b2a8f2a42266e67540R200-R203) [[2]](diffhunk://#diff-ef0c46a534c74a4d5c6594bd14a6cf982c51f861ca18e8b2a8f2a42266e67540R289-R291) [[3]](diffhunk://#diff-ef0c46a534c74a4d5c6594bd14a6cf982c51f861ca18e8b2a8f2a42266e67540R379-R380)
* [`heat/core/arithmetics.py`](diffhunk://#diff-6eb65c3e0040c50b6ac4c6f621b37b8e74eabfdcfe52b0e16fc0cf6c413b924aL1625-R1642): Added checks for unsupported int64 and float64 operations on MPS in functions like `hypot` and `hypot_`. [[1]](diffhunk://#diff-6eb65c3e0040c50b6ac4c6f621b37b8e74eabfdcfe52b0e16fc0cf6c413b924aL1625-R1642) [[2]](diffhunk://#diff-6eb65c3e0040c50b6ac4c6f621b37b8e74eabfdcfe52b0e16fc0cf6c413b924aR1675-R1690) [[3]](diffhunk://#diff-6eb65c3e0040c50b6ac4c6f621b37b8e74eabfdcfe52b0e16fc0cf6c413b924aR1733-R1751)
* [`heat/core/dndarray.py`](diffhunk://#diff-88beaa741e8d62f4e9ee96ba7e05ab8811a7c5a05422c50e1f0dc78a3a56e163L211-R220): Updated size calculation and type casting to handle unsupported double precision and complex types on MPS. [[1]](diffhunk://#diff-88beaa741e8d62f4e9ee96ba7e05ab8811a7c5a05422c50e1f0dc78a3a56e163L211-R220) [[2]](diffhunk://#diff-88beaa741e8d62f4e9ee96ba7e05ab8811a7c5a05422c50e1f0dc78a3a56e163R501-R515)

These changes ensure that the codebase is compatible with MPS, providing better support for Apple devices while maintaining functionality across different platforms.












## Reference

- Current status of [PyTorch's MPS operations coverage](https://github.com/pytorch/pytorch/issues/77764)



Issue/s resolved: #1053 

## Changes proposed:
-
-
-
-

## Type of change
<!--
i.e.
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Documentation update
--->

## Memory requirements
<!--- Compare memory requirements to previous implementation / relevant torch operations if applicable:
- in distributed and non-distributed mode
- with `split=None` and `split not None`

This can be done using https://github.com/pythonprofilers/memory_profiler for CPU memory measurements,
GPU measuremens can be done with https://pytorch.org/docs/master/generated/torch.cuda.max_memory_allocated.html.
These tools only profile the memory used by each process, not the entire function.
--->

## Performance
<!--- Compare performance to previous implementation / relevant torch operations if applicable:
- in distributed and non-distributed mode
- with `split=None` and `split not None`

Python has an embedded profiler: https://docs.python.org/3.9/library/profile.html
Again, this will only profile the performance on each process. Printing the results with many processes
my be illegible. It may be easiest to save the output of each to a file.
--->


## Due Diligence
- [ ] ~~All split configurations tested~~ does not apply
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Title of PR is suitable for corresponding CHANGELOG entry

#### Does this change modify the behaviour of other functions? If so, which?
no

mrfh92 commented on heat/core/_operations.py:
```diff
@@ -173,6 +173,10 @@ def __get_out_params(target, other=None, map=None):
         sanitation.sanitize_out(out, output_shape, output_split, output_device, output_comm)
         t1, t2 = sanitation.sanitize_distribution(t1, t2, target=out)
 
+    # MPS does not support float32
```
do you mean: "MPS does not support float64?"

ClaudiaComito commented:
@mrfh92 this is too unstable to release now, esp. with me being the only one who can test things and being away for the next few weeks. Let's leave this for later

mrfh92 commented:
Tests failed; I rerun them to check whether thats just a HW problem.

ClaudiaComito commented:
> Tests failed; I rerun them to check whether thats just a HW problem.

Thanks, sadly it looks like an actual problem, conveniently without error message. I'll debug it, will probably get to it next week.

JuanPedroGHM commented on heat/core/relational.py:
```diff
@@ -165,6 +166,9 @@ def equal(x: Union[DNDarray, float, int], y: Union[DNDarray, float, int]) -> boo
                 y = y.balance()
 
     result_type = types.result_type(x, y)
+    is_mps = x.larray.is_mps or y.larray.is_mps
+    if is_mps and result_type is types.float64:
+        result_type = types.float32
```
Instead of checking every time after calling ```types.result_type```, the check could be done inside ```types.result_type()```. This would save a lot of extra if statements and less chance of possibly forgetting to add that.

JuanPedroGHM commented on heat/core/signal.py:
```diff
@@ -116,6 +116,10 @@ def convolve(a: DNDarray, v: DNDarray, mode: str = "full") -> DNDarray:
         except TypeError:
             raise TypeError(f"non-supported type for filter: {type(v)}")
     promoted_type = promote_types(a.dtype, v.dtype)
+    if a.larray.is_mps and promoted_type == float64:
+        # cannot cast to float64 on MPS
+        promoted_type = float32
+
```
Same with ```promote_types```.

ClaudiaComito commented:
New errors after merging `main`:

- [ ] `test_det` (`NotImplementedError`): `heat/core/linalg/tests/test_basics.py:98`
    *   Error Message: `NotImplementedError: The operator 'aten::_linalg_det.result' is not currently implemented for the MPS device...`

- [ ] `test_inv` (`AttributeError`): `heat/core/linalg/tests/test_basics.py:291`
    *   Error Message: `AttributeError: module 'heat' has no attribute 'floa32'`

- [ ] `test_batched_qr_split0` (`NotImplementedError`): `heat/core/linalg/tests/test_qr.py:161`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...`

- [ ] `test_batched_qr_splitNone` (`NotImplementedError`): `heat/core/linalg/tests/test_qr.py:138`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...`

- [ ] `test_qr_split0` (`NotImplementedError`): `heat/core/linalg/tests/test_qr.py:92`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...`

- [ ] `test_qr_split1orNone` (`NotImplementedError`): `heat/core/linalg/tests/test_qr.py:28`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...`

- [ ] `test_singvals_only` (`NotImplementedError`): `heat/core/linalg/tests/test_svd.py:72`
    *   Error Message: `NotImplementedError: The operator 'aten::_linalg_svd.U' is not currently implemented for the MPS device...`

- [ ] `test_hsvd_rank_part2` (`NotImplementedError`): `heat/core/linalg/tests/test_svdtools.py:163`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `random_orthogonal`)

- [ ] `test_isvd` (`NotImplementedError`): `heat/core/linalg/tests/test_svdtools.py:288`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `random_orthogonal`)

- [ ] `test_rsvd` (`NotImplementedError`): `heat/core/linalg/tests/test_svdtools.py:232`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `rsvd`)

- [ ] `test_gcd` (`NotImplementedError`): `heat/core/arithmetics.py:1570`
    *   Error Message: `NotImplementedError: The operator 'aten::gcd.out' is not currently implemented for the MPS device...`

- [ ] `test_gcd` (`TypeError`): `heat/core/tests/test_arithmetics.py:2282`
    *   Error Message: `TypeError: Expected integer input, got <class 'heat.core.types.int64'>, <class 'heat.core.types.int64'>`

- [ ] `test_gcd_` (`NotImplementedError`): `heat/core/arithmetics.py:1628`
    *   Error Message: `NotImplementedError: The operator 'aten::gcd.out' is not currently implemented for the MPS device...`

- [ ] `test_gcd_` (`AttributeError`): `heat/core/tests/test_arithmetics.py:2310`
    *   Error Message: `AttributeError: 'int' object has no attribute 'shape'`

- [ ] `test_lcm` (`NotImplementedError`): `heat/core/arithmetics.py:1883`
    *   Error Message: `NotImplementedError: The operator 'aten::lcm.out' is not currently implemented for the MPS device...`

- [ ] `test_lcm` (`TypeError`): `heat/core/tests/test_arithmetics.py:2712`
    *   Error Message: `TypeError: Expected integer input, got <class 'heat.core.types.int64'>, <class 'heat.core.types.int64'>`

- [ ] `test_lcm_` (`NotImplementedError`): `heat/core/arithmetics.py:1941`
    *   Error Message: `NotImplementedError: The operator 'aten::lcm.out' is not currently implemented for the MPS device...`

- [ ] `test_lcm_` (`AttributeError`): `heat/core/tests/test_arithmetics.py:2740`
    *   Error Message: `AttributeError: 'int' object has no attribute 'shape'`

- [ ] `test_arange` (`TypeError`): `heat/core/tests/test_factories.py:110`
    *   Error Message: `TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`

- [ ] `test_array` (`TypeError`): `heat/core/tests/test_factories.py:149`
    *   Error Message: `TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`

- [ ] `test_linspace` (`TypeError`): `heat/core/tests/test_factories.py:746`
    *   Error Message: `TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`

- [ ] `test_load_npy_float` (`TypeError`): `heat/core/tests/test_io.py:797`
    *   Error Message: `TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`

- [ ] `test_load_npy_int` (`ValueError`): `heat/core/tests/test_io.py:773`
    *   Error Message: `ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 4 and the array at index 1 has size 1`

- [ ] `test_isneginf` (`NotImplementedError`): `heat/core/tests/test_logical.py:405`
    *   Error Message: `NotImplementedError: The operator 'aten::isneginf.out' is not currently implemented for the MPS device...`

- [ ] `test_isposinf` (`NotImplementedError`): `heat/core/tests/test_logical.py:439`
    *   Error Message: `NotImplementedError: The operator 'aten::isposinf.out' is not currently implemented for the MPS device...`

- [ ] `test_unique` (`NotImplementedError`): `heat/core/tests/test_manipulations.py:3613`
    *   Error Message: `NotImplementedError: The operator 'aten::unique_dim' is not currently implemented for the MPS device...`

- [ ] `test_abs` (`TypeError`): `heat/core/tests/test_rounding.py:23`
    *   Error Message: `TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`

- [ ] `test_round` (`TypeError`): `heat/core/tests/test_rounding.py:281`
    *   Error Message: `TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`

- [ ] `test_vmap` (`NotImplementedError`): `heat/core/tests/test_vmap.py:32`
    *   Error Message: `NotImplementedError: The operator 'aten::_linalg_svd.U' is not currently implemented for the MPS device...`

- [ ] `test_vmap_with_chunks` (`NotImplementedError`): `heat/core/tests/test_vmap.py:94`
    *   Error Message: `NotImplementedError: The operator 'aten::_linalg_svd.U' is not currently implemented for the MPS device...`

- [ ] `test_incrementalpca_full_rank_reached_split0` (`NotImplementedError`): `heat/decomposition/tests/test_pca.py:268`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `_isvd`)

- [ ] `test_incrementalpca_truncation_happens_split1` (`NotImplementedError`): `heat/decomposition/tests/test_pca.py:312`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `_isvd`)

- [ ] `test_random_known_rank` (`NotImplementedError`): `heat/utils/data/tests/test_matrixgallery.py:106`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `random_orthogonal`)

- [ ] `test_random_known_singularvalues` (`NotImplementedError`): `heat/utils/data/tests/test_matrixgallery.py:86`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `random_orthogonal`)

- [ ] `test_random_orthogonal` (`NotImplementedError`): `heat/utils/data/tests/test_matrixgallery.py:69`
    *   Error Message: `NotImplementedError: The operator 'aten::linalg_qr.out' is not currently implemented for the MPS device...` (Caused by `qr` call within `random_orthogonal`)

ClaudiaComito commented on heat/core/linalg/tests/test_qr.py:
```diff
@@ -132,18 +140,20 @@ def test_batched_qr_splitNone(self):
-            ht.allclose(q.transpose([0, 1, 3, 2]) @ q, batched_id, atol=1e-6, rtol=1e-6)
-        )
-        self.assertTrue(ht.allclose(q @ r, x, atol=1e-6, rtol=1e-6))
+        # skip float64 tests on MPS
```
@JuanPedroGHM we only want to skip a portion of a test function, I'm not sure how to do that with a unittest.skipIf decorator

ClaudiaComito commented on heat/core/linalg/solver.py:
```diff
@@ -339,7 +339,7 @@ def solve_triangular(A: DNDarray, b: DNDarray) -> DNDarray:
             b_lshapes_cum = torch.hstack(
                 [
-                    torch.zeros(1, dtype=torch.int32, device=tdev),
+                    torch.zeros(1, dtype=torch.int64, device=tdev),
```
We should use int64 for indices so we don't run into problems with very large arrays. In the beginning we weren't paying much attention to this, so we still have a few int32 relics.

ClaudiaComito commented:
@JuanPedroGHM thanks for reviewing, I've addressed all your points I think, refined the FFT and complex numbers tests after upgrading to Mac OS 15.  

@mtar thanks for the new runner. I cannot reproduce the error locally. I noticed the tests run on Python 3.9 instead of 3.11, not sure if that makes any difference.

mtar commented:
sorry about that, next time 3.11 is used throughout the run.

ClaudiaComito commented:
@JuanPedroGHM @mtar  I ran the tests on my laptop with Mac OSX 15.3.1, python 3.11, Pytorch 2.6,  and I cannot reproduce the error.
